5117_Sprint/09-Haoran/Chapter4/src/main/java/PCounting.java

```diff
@@ -0,0 +1,29 @@
+import java.util.Arrays;
+import java.util.stream.IntStream;
+
+public class PCounting {
+
+    public static double pCount(IntStream stream, int k){
+        int numBuckets = (int) Math.pow(2,k);
+        int[] maxZeros = new int[numBuckets];
+        Arrays.fill(maxZeros, 0);
+        stream.map(Integer::hashCode)
+                .forEach(l -> {if (maxZeros[l & (numBuckets - 1)] < trailingZeroes(l >> k))
+                    maxZeros[(int)l & (numBuckets - 1)] = trailingZeroes(l >> k);});
+
+        return Math.pow(2, IntStream.of(maxZeros).sum()/numBuckets) *numBuckets *0.79402;
+    }
+
+    private static int trailingZeroes(int hashNum){
```

Comment:
This function is only used inside this class. I see no reason for it to be static.